TIBERIANDAWN/INFANTRY.CPP

```diff
@@ -2786,15 +2810,15 @@ RadioMessageType InfantryClass::Receive_Message(RadioClass * from, RadioMessageT
 		**	Just received a kick! Take some damage.
 		*/
 		case	RADIO_KICK:
-			damage = Infantry_Kick_Damage[Random_Pick(0, (int)(sizeof(Infantry_Kick_Damage) / sizeof(Infantry_Kick_Damage[0])))];
+			damage = Infantry_Kick_Damage[Random_Pick(0, (int)(sizeof(Infantry_Kick_Damage) / sizeof(Infantry_Kick_Damage[0])) - 1)];
```

Comment:
A classic off-by-one error. I wonder how many of those invalid reads are in the code.
Maybe create a macro/inline template to properly access such arrays?

```cpp
#define MAX_ARRAY_INDEX(array) (int)(sizeof(array) / sizeof(array[0])) - 1
```

I out of C++ for several years but I am sure an inline template function could make this even more secure.

Reply:
> A classic off-by-one error. I wonder how many of those invalid reads are in the code.
> Maybe create a macro/inline template to properly access such arrays?
> 
> ```c++
> #define MAX_ARRAY_INDEX(array) (int)(sizeof(array) / sizeof(array[0])) - 1
> ```
> 
> I out of C++ for several years but I am sure an inline template function could make this even more secure.

You probably want to wrap that inside another pair of parentheses in case you ever want to use an operator with a higher priority after using this macro. It is a little bit nitpicky and I don't see a real use of this happening in this project, but I prefer to be extra cautious when using macros.

Instead of:
`#define MAX_ARRAY_INDEX(array) (int)(sizeof(array) / sizeof(array[0])) - 1`
Use:
`#define MAX_ARRAY_INDEX(array) ((int)(sizeof(array) / sizeof(array[0])) - 1)`

A situation like this could return undesired side-effects (imagine an array size of 5, thus max index 4):
`int a = MAX_ARRAY_INDEX(array) * 2;`  The intended value of 'a' is 8.
Would be understood as:
`int a = (int)(sizeof(array) / sizeof(array[0])) - 1 * 2;`  The actual value of 'a' is 3.

Reply:
Correct, my code was made up on the fly and I mentioned there is probably a more typesafe way in C++ to do that.